File: src/dp/cut_rod.cpp
// cut_rod.cpp — CLRS 4e §14.1：naive（計呼叫數）vs bottom-up
#include <iostream>
#include <vector>

long long calls = 0;

int cutRodNaive(const std::vector<int>& p, int n) {
    ++calls;
    if (n == 0) return 0;
    int q = -1;
    for (int i = 1; i <= n; ++i)
        q = std::max(q, p[i] + cutRodNaive(p, n - i));   // 第一刀切 i，其餘遞迴
    return q;
}

std::pair<int, std::vector<int>> bottomUpCutRod(const std::vector<int>& p, int n) {
    std::vector<int> r(n + 1, 0), s(n + 1, 0);   // r[j] = 長度 j 的最佳收益；s[j] = 第一刀
    for (int j = 1; j <= n; ++j) {
        int q = -1;
        for (int i = 1; i <= j; ++i)
            if (p[i] + r[j - i] > q) { q = p[i] + r[j - i]; s[j] = i; }
        r[j] = q;
    }
    return {r[n], s};
}

int main() {
    // CLRS 圖 14.1 的價格表：p[i] = 長度 i 的售價
    std::vector<int> p = {0, 1, 5, 8, 9, 10, 17, 17, 20, 24, 30};
    int n = 10;
    int naive = cutRodNaive(p, n);
    auto [best, s] = bottomUpCutRod(p, n);
    std::cout << "naive = " << naive << "（遞迴呼叫 " << calls << " 次 = 2^n）\n";
    std::cout << "bottom-up = " << best << "，切法：";
    while (n > 0) { std::cout << s[n] << ' '; n -= s[n]; }
    std::cout << '\n';
    return 0;
}

File: src/dp/cut_rod.py
# 改寫自官方 Chapter 14/cut_rod.py（MIT License）：bottom-up＋memoized 兩版全文
p = [0, 1, 5, 8, 9, 10, 17, 17, 20, 24, 30]    # CLRS 圖 14.1 價目表（p[0] 不用）


def bottom_up_cut_rod(p, n):
    r = [0] * (n + 1)
    s = [0] * (n + 1)                     # s[j] = 長度 j 的最佳第一刀
    for j in range(1, n + 1):             # 子問題由小到大
        q = float("-inf")
        for i in range(1, j + 1):
            if p[i] + r[j - i] > q:
                q = p[i] + r[j - i]
                s[j] = i
        r[j] = q
    return r, s


def memoized_cut_rod(p, n, r=None):       # top-down：算過就查表
    if r is None:
        r = [None] * (n + 1)
    if r[n] is not None:
        return r[n]
    if n == 0:
        q = 0
    else:
        q = float("-inf")
        for i in range(1, n + 1):
            q = max(q, p[i] + memoized_cut_rod(p, n - i, r))
    r[n] = q
    return q


r, s = bottom_up_cut_rod(p, 10)
print("bottom-up r[10] =", r[10])
cuts, j = [], 10
while j > 0:                              # 由 s 表回溯切法
    cuts.append(s[j])
    j -= s[j]
print("切法：", cuts)
print("memoized r(10) =", memoized_cut_rod(p, 10))
print("r 表：", r)
